rp2040-hal/src/uart.rs

```diff
@@ -1,3 +1,473 @@
-//! Universal asynchronous receiver-transmitter (UART)
+//! Universal Asynchronous Receiver Transmitter (UART)
 // See [Chapter 4 Section 2](https://datasheets.raspberrypi.org/rp2040/rp2040_datasheet.pdf) for more details
-// TODO
+
+use core::convert::Infallible;
+use core::ops::Deref;
+use embedded_time::fixed_point::FixedPoint;
+use embedded_time::rate::Baud;
+use embedded_time::rate::Hertz;
+
+use embedded_hal::serial::{Read, Write};
+
+use nb::Error::{Other, WouldBlock};
+
+use crate::pac::{
+    uart0::{uartlcr_h::W as UART_LCR_H_Writer, RegisterBlock},
+    UART0, UART1,
+};
+
+/// Error type for UART operations.
+#[derive(Debug)]
+pub enum Error {
+    /// Bad argument : when things overflow, ...
+    BadArgument,
+}
+
+/// When there's a read error.
+pub struct ReadError<'err> {
+    /// The type of error
+    pub err_type: ReadErrorType,
+
+    /// Reference to the data that was read but eventually discared because of the error.
+    pub discared: &'err [u8],
+}
+
+/// Possible types of read errors. See Chapter 4, Section 2 §8 - Table 436: "UARTDR Register"
+pub enum ReadErrorType {
+    /// Triggered when the FIFO (or shift-register) is overflowed.
+    Overrun,
+
+    /// Triggered when a break is received
+    Break,
+
+    /// Triggered when there is a parity mismatch between what's received and our settings.
+    Parity,
+
+    /// Triggered when the received character didn't have a valid stop bit.
+    Framing,
+}
+
+/// State of the UART Peripheral.
+pub trait State {}
+
+/// Trait to handle both underlying devices (UART0 & UART1)
+pub trait UARTDevice: Deref<Target = RegisterBlock> {}
+
+impl UARTDevice for UART0 {}
+impl UARTDevice for UART1 {}
+
+/// UART is enabled.
+pub struct Enabled;
+
+/// UART is disabled.
+pub struct Disabled;
+
+impl State for Enabled {}
+impl State for Disabled {}
+
+/// Data bits
+pub enum DataBits {
+    /// 5 bits
+    Five,
```

Comment:
I think you're going a bit overboard with doc strings here.
Reader can probably infer that DataBits are Data Bits, and that Five represents 5 bits

Reply:
Totally agreeing with you but I think it wouldn't compile without the comments. Will check again !

Reply:
```
error: missing documentation for a variant
  --> rp2040-hal/src/uart.rs:70:5
   |
70 |     Five,
   |     ^^^^
   |
note: the lint level is defined here
  --> rp2040-hal/src/lib.rs:6:9
   |
6  | #![deny(missing_docs)]
   |         ^^^^^^^^^^^^

error: aborting due to previous error

error: could not compile `rp2040-hal`

```

Reply:
This was bad advice, my apologies.
I don't like it but it's better than writing manual exceptions to doc requirements.